src/simularium/VisData.ts

```diff
@@ -98,91 +53,58 @@ class VisData {
         if (util.ThreadUtil.browserSupportsWebWorkers()) {
             this.setupWebWorker();
         }
-        this.frameCache = [];
-        this.frameDataCache = [];
-        this.cacheFrame = -1;
-        this.enableCache = true;
+        this.currentFrameNumber = -1;
+        this.frameCache = new VisDataCache();
         this.frameToWaitFor = 0;
         this.lockedForFrame = false;
         this.timeStepSize = 0;
+
+        this.onError = noop;
     }
 
-    //get time() { return this.cacheFrame < this.frameDataCache.length ? this.frameDataCache[this.cacheFrame] : -1 }
-    public get currentFrameData(): FrameData {
-        if (this.frameDataCache.length > 0) {
-            if (this.cacheFrame < 0) {
-                return this.frameDataCache[0];
-            } else if (this.cacheFrame >= this.frameDataCache.length) {
-                return this.frameDataCache[this.frameDataCache.length - 1];
-            } else {
-                return this.frameDataCache[this.cacheFrame];
-            }
-        }
+    public setOnError(onError: (error: FrontEndError) => void): void {
+        this.onError = onError;
+    }
 
-        return { frameNumber: 0, time: 0 };
+    public get currentFrameData(): CachedFrame {
+        let currentData: CachedFrame = nullCachedFrame();
+        if (!this.frameCache.hasFrames()) {
+            return currentData;
+        }
+        if (this.currentFrameNumber < 0) {
+            currentData = this.frameCache.getFirstFrame();
+            this.currentFrameNumber = currentData.frameNumber;
+        } else if (
+            this.frameCache.containsFrameAtFrameNumber(this.currentFrameNumber)
+        ) {
+            currentData = this.frameCache.getFrameAtFrameNumber(
+                this.currentFrameNumber
+            );
+        }
+        return currentData;
     }
```

Comment:
A few observations here:

1. Nit: since the function immediately knows its return value when any of the `if`s pass, it's not clear that you gain anything by keeping `currentData` around rather than just early-returning and inlining `return nullCachedFrame();` as the default case at the end.
2. This function makes a decent argument for `frameNumber` as the primary key of `VisDataCache`. It's possible that in practice the cache is always small enough to be efficiently traversable, but in principle a developer might be surprised that using this `get` accessor - without `()` after it or anything - triggers an O(n) operation!
3. In fact, it currently triggers two: `containsFrameAtFrameNumber` and `getFrameAtFrameNumber`, which redundantly find the same entry (or discover that it doesn't exist) twice. I'd suggest instead that `getFrameAtFrameNumber` should just handle misses without erroring, and its return type should be expanded to `CacheEntry | undefined`. Then the check above can look more like this:

```typescript
const frame = this.frameCache.getFrameAtFrameNumber(this.currentFrameNumber);
if (frame !== undefined) {
    return frame;
}
```

Mini-soapbox: in general, I'm extremely wary of throwing errors in non-catastrophic circumstances, and instead try to include the failure case in a function's return type. If the type signature of a function is a contract, errors represent a "leak" in the contract. E.g. this method already doesn't just return a `CachedFrame` - it returns a `CachedFrame` *or* it errors out - but now TypeScript doesn't know to enforce handling that case. (Note that this is far from a widely shared opinion, or errors would work differently at the language level!)

Reply:
On deciding how serious the error is: what are possible scenarios where we might actually request a frame number that can't be found?  Are they indicative of bad data or just data that hasn't arrived yet? Could it happen in normal usage of the viewer, like moving the time slider way ahead of what's in cache?

Reply:
Also this is a great point about calling both containsFrameAtNumber and getFrameAtNumber.. lookup into the data structure twice (ie walking the list twice) does seem wasteful

Reply:
1. Ok, so I like this. I'll return `nullCachedFrame()` if `hasFrames` fails,  as well as at the end.
2. I'm going to focus on addressing your more immediate concerns and then circling back to these deeper questions about Maps/ring buffers/linked lists and optimizing the cache data structure.
3. Yes to your snippet and not calling both those methods, and yes to adding undefined to the return type of `getFrameAtTime` and `getFrameAtFrameNumber`.
But what about the errors in `getFrameAtCondition`, what do you think should happen if a bad argument gets passed in? Console warnings? These methods can be called via the controller API so in theory a valid trajectory could be loaded and wacky arguments could get passed in.

Reply:
Ok so I'm coming along for this ride. Pretty much removing all error handling from `VisDataCache` on the assumption that if you are a function who tried to get a frame and received `undefined` it's your problem to decide what to do.